DOC: update the pandas.DataFrame.plot.area docstring

DZPM opened the pull request:
- [x] PR title is "DOC: update the pandas.DataFrame.plot.area docstring"
- [x] The validation script passes: `scripts/validate_docstrings.py pandas.DataFrame.plot.area`
- [x] The PEP8 style check passes: `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] The html version looks good: `python doc/make.py --single pandas.DataFrame.plot.area`
- [x] It has been proofread on language by another sprint participant

```
################################################################################
#################### Docstring (pandas.DataFrame.plot.area) ####################
################################################################################

Draw an area plot of the input series using matplotlib.

An area plot displays graphically quantitative data.
This function wraps the matplotlib area function.

Parameters
----------
x : label or position, optional
    Coordinates for X point.
y : label or position, optional
    Coordinates for Y point.
kwds : optional
    Keyword arguments to pass on to :py:meth:`pandas.DataFrame.plot`.

Returns
-------
matplotlib.AxesSubplot or numpy.ndarray
    Area plot generated.

See Also
--------
pandas.DataFrame.plot : Draw a plot using matplotlib.

Examples
--------
.. plot::
    :context: close-figs

    This example draws an area plot based on the length and width of
    some animals, displayed in three bins

    >>> df = pd.DataFrame({
    ...     'sales': [3, 2, 3, 9, 10, 6],
    ...     'visits': [20, 42, 28, 62, 81, 50],
    ... })
    >>> area = df.plot.area()

################################################################################
################################## Validation ##################################
################################################################################

Docstring for "pandas.DataFrame.plot.area" correct. :)
```

DZPM committed:
DOC: Improve docstring for pandas.DataFrame.plot.area

kynan commented on pandas/plotting/_core.py:
```diff
@@ -2818,18 +2818,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
-        Area plot
+        Draw an area plot of the input series using matplotlib.
+
+        An area plot displays graphically quantitative data.
```
displays quantitative data visually

dukebody replied:
It would be great to add some info about why this function is useful or some basic use cases. `Displays quantitative data visually` doesn't really give me very interesting information about what is the shape or usage of the plot.

kynan commented on pandas/plotting/_core.py:
```diff
@@ -2818,18 +2818,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
-            Coordinates for each point.
-        `**kwds` : optional
+        x : label or position, optional
+            Coordinates for X point.
```
X axis

dukebody replied:
From this description of the arguments `x` and `y` I really don't understand what they change in the plot. Can we make the parameter description a bit longer explaining what do they mean?

kynan commented on pandas/plotting/_core.py:
```diff
@@ -2818,18 +2818,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
+        x : label or position, optional
+            Coordinates for X point.
+        y : label or position, optional
+            Coordinates for Y point.
```
Y axis

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -2818,18 +2818,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
+
+        See Also
+        --------
+        pandas.DataFrame.plot : Draw a plot using matplotlib.
```
The explanation here is a bit strange, as it would seem to say that this one is not plotting with matplotlib

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -2818,18 +2818,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
         -------
-        axes : matplotlib.AxesSubplot or np.array of them
+        matplotlib.AxesSubplot or numpy.ndarray
+            Area plot generated.
```
Can you explain something about when ndarray is returned?

dukebody replied:
Yep. In this case the existing explanation `Area plot generated` doesn't tell a lot. How is a `numpy.ndarray` the area plot generated?

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -2818,18 +2818,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
     def area(self, x=None, y=None, **kwds):
         """
-        Area plot
+        Draw an area plot of the input series using matplotlib.
```
"input series" -> input is a bit strange wording as it is the calling object (not passed as input) + it should be DataFrame instead of Series ?

dukebody replied:
I don't know if we want to mention the underlying library this is using at this level. Does the user really care so much? In the extended description is more OK IMO.

dukebody replied:
Also, as @jorisvandenbossche says, the description doesn't make sense for dataframes. Probably what the function does is to display a stacked area plot?

jorisvandenbossche reviewed (commented):
Thanks! added some comments

dukebody commented:
Can you post a screenshot of the rendered HTML? It will help reviewing without having to check out your branch and compiling it.

dukebody commented on pandas/plotting/_core.py:
```diff
@@ -2818,18 +2818,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
-        x, y : label or position, optional
-            Coordinates for each point.
-        `**kwds` : optional
+        x : label or position, optional
```
This should follow the guide: https://python-sprints.github.io/pandas/guide/pandas_docstring.html#section-3-parameters

So it should indicate the type of the parameter in the first line (plus optional), and later the meaning of the parameter.

dukebody commented on pandas/plotting/_core.py:
```diff
@@ -2818,18 +2818,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
+            Coordinates for X point.
+        y : label or position, optional
+            Coordinates for Y point.
+        kwds : optional
```
As mentioned in other PRs, you should use `**kwds` instead. Sorry for the great big confusion here.

DZPM replied:
Using `**kwds` raises errors in the current version of the validator:

	Errors in parameters section
		Parameters {'kwds'} not documented
		Unknown parameters {'**kwds'}

I think it's important to be consistent between the validator and the code.

Of course, once the validator is updated, I agree this must be changed.

jorisvandenbossche replied:
You can ignore that error in the validation script for now.

DZPM replied:
Ok, updated!

dukebody commented on pandas/plotting/_core.py:
```diff
@@ -2818,18 +2818,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
+        .. plot::
+            :context: close-figs
+
+            This example draws an area plot based on the length and width of
```
1) I think is better to put the text before the start of the `plot` block, and leave the block only for the code that constructs the plot.
2) The example this sentence talks about doesn't correspond to the dataframe below.
3) Don't write in 3rd person `This example does this`. Better write directly in imperative or present form, like `Draw an area plot based on` or `We can draw an area plot based on...`, but IMO mentioning `this example` is somewhat redundant when we are already in the examples section. :)

dukebody commented on pandas/plotting/_core.py:
```diff
@@ -2818,18 +2818,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
+            :context: close-figs
+
+            This example draws an area plot based on the length and width of
+            some animals, displayed in three bins
```
In three bins? Where does the three come from?

DZPM replied:
Yes, it's wrong, copied it from another example. Removed it.

dukebody commented on pandas/plotting/_core.py:
```diff
@@ -2818,18 +2818,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
+            ...     'sales': [3, 2, 3, 9, 10, 6],
+            ...     'visits': [20, 42, 28, 62, 81, 50],
+            ... })
+            >>> area = df.plot.area()
```
I'm not sure about this, but if the result of this method is an axis object, perhaps it's better to assign it to a variable named `ax` like in other examples.

Can you add other examples where the usage of the `x` and `y` parameters is shown? Also an example where a `numpy.ndarray` is returned.

DZPM replied:
Not sure about how to define a "usable" example, a little help would be appreciated!

dukebody reviewed (changes requested):
I believe this needs some better explanations of the extended description, arguments and more examples that help to understand how this graph works.

dukebody commented:
Is there anything we can do to help finishing the work on this PR?

DZPM commented:
I'll update the PR later this week, just need some time :+1:

DZPM commented:
Fixed most of the suggestions, except for:
- writing `**kwds` instead of `kwds`. That would break the validation.
- more/better examples: I ask for inspiration :)

Also, a screenshot of the example:
![image](https://user-images.githubusercontent.com/353447/37460194-ccceaa6c-284a-11e8-8a43-863b4a1131e1.png)

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -3115,19 +3115,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
     def area(self, x=None, y=None, **kwds):
         """
-        Area plot
+        Draw an stacked area plot.
```
an -> a

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -3115,19 +3115,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
-        x, y : label or position, optional
-            Coordinates for each point.
-        `**kwds` : optional
+        x : label or position, optional, default None
```
you can leave out the ", default None" part (saying it is "optional" should be informative enough"). 

But we should specify here what it used for `x` by default. I suppose this is the index?

(and the same comment for `y`

TomAugspurger replied:
Yes I believe it's the index by default.

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -3115,19 +3115,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
+            Coordinates for X axis.
+        y : label or position, optional, default None
+            Coordinates for Y axis.
+        **kwds : optional
```
Can you also document the `stacked` keyword? I know it is not in the signature, but it is a relevant one for this type of plot, so I would add it. 
See http://pandas.pydata.org/pandas-docs/stable/visualization.html#area-plot

(the validation script will complain about it, but you can ignore that again)

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -3115,19 +3115,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
         -------
-        axes : :class:`matplotlib.axes.Axes` or numpy.ndarray of them
+        matplotlib.axes.Axes or numpy.ndarray
+            Area plot, or array of area plots.
```
It would be good to add when an array is returned, which is when `subplots=True` is specified

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -3115,19 +3115,42 @@ def kde(self, bw_method=None, ind=None, **kwds):
+            ...     'sales': [3, 2, 3, 9, 10, 6],
+            ...     'visits': [20, 42, 28, 62, 81, 50],
+            ... })
+            >>> ax = df.plot.area()
```
Some ideas for the examples:

- I would also show one where you use the `x` and `y` keywords
- I would show one with `stacked=False`
- I think it is typical to use area plots with time series data. So maybe it would be nice to add a DatetimeIndex to the `df`.

jorisvandenbossche reviewed (commented):
Thanks for the updates! Added some more comments

TomAugspurger commented:
> writing **kwds instead of kwds. That would break the validation.

You can ignore this, `**kwargs` is correct, the script is incorrect :)

TomAugspurger commented:
http://pandas.pydata.org/pandas-docs/stable/visualization.html#area-plot has an example. The same thing with a datetimeidnex from `pd.date_range` would be nice (maybe a bit wider of an aspect ratio).

DZPM commented:
Hi,

I've updated with all the proposals, and rebased.

Screenshots of the rendered example section:
![image](https://user-images.githubusercontent.com/353447/37930583-7eae75a4-3143-11e8-8319-ec24aac75a34.png)
![image](https://user-images.githubusercontent.com/353447/37930595-856efb52-3143-11e8-8a47-3f55c47f19e4.png)
![image](https://user-images.githubusercontent.com/353447/37930606-8c1a49f2-3143-11e8-9cc4-55f39b1627df.png)
![image](https://user-images.githubusercontent.com/353447/37930619-913c5182-3143-11e8-8f91-ec5f1ec82ca0.png)
![image](https://user-images.githubusercontent.com/353447/37930629-97613dd4-3143-11e8-9f39-e1c7ec03d63d.png)

It looks much better :) Thanks for the feedback!

DZPM commented:
Ping @kynan @jorisvandenbossche @TomAugspurger @dukebody 

Feel free to keep reviewing, or do the merge :)

jorisvandenbossche commented:
@DZPM Thanks for the reminder! 
Can you add a datetime index to the example frame? See https://github.com/pandas-dev/pandas/pull/20170#issuecomment-373364136

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -3188,19 +3188,85 @@ def kde(self, bw_method=None, ind=None, **kwds):
-            Coordinates for each point.
-        `**kwds` : optional
+        x : label or position, optional
+            Index, coordinates for X axis.
```
The "Index, " part is a bit confusing I think. I would reword this as : `Coordinates for the Y axis. By default uses the index.`

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -3188,19 +3188,85 @@ def kde(self, bw_method=None, ind=None, **kwds):
+        x : label or position, optional
+            Index, coordinates for X axis.
+        y : label or position, optional
+            Index, coordinates for Y axis.
```
Here the default is not the index, but all columns. So maybe something like "Column to plot. By default uses all columns." ?

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -3188,19 +3188,85 @@ def kde(self, bw_method=None, ind=None, **kwds):
+            ...     'visits': [20, 42, 28, 62, 81, 50],
+            ... })
+            >>> ax = df.plot.area(y='sales')
+            >>> ax = df.plot.area(y='signups')
```
I think showing only the first is fine to illustrate the `y` keyword (to save some vertical screen estate)

(if you want to have all of them on a separate figure, one could also use `subplots=True`)

jorisvandenbossche reviewed (commented):
Added a few more comments

BTW, for future reference, if you add new commits to the branch with additional changes instead of squashing everything into a single commit, it is a bit easier to see what has been updated since the last review

datapythonista commented:
@DZPM can you please update this PR based on the previous feedback

DZPM commented:
Sure, will do. (Sorry, I've been busy and the PR slipped my mind...)

DZPM committed:
Improved with suggestions

DZPM commented:
Done. Added the changes in a new commit, and rebased.

datapythonista commented on pandas/plotting/_core.py:
```diff
@@ -3337,19 +3337,87 @@ def kde(self, bw_method=None, ind=None, **kwds):
+            ...     'signups': [5, 5, 6, 12, 14, 13],
+            ...     'visits': [20, 42, 28, 62, 81, 50],
+            ... }, index=pd.date_range(start='2018/01/01', end='2018/07/01',
+            ...                        freq='M'))
```
is the `.. plot::` directive resetting the namespace? Otherwise, It'd be good to reuse the DataFrame instead of repeating it in every example.

DZPM replied:
Had some issues with the namespace, so I kept it in different places.

Don't have the full environment right now, can't check if it also works the other way.

jorisvandenbossche replied:
Recreating the dataframe each time should not be needed

datapythonista commented on pandas/plotting/_core.py:
```diff
@@ -3337,19 +3337,87 @@ def kde(self, bw_method=None, ind=None, **kwds):
+
+        See Also
+        --------
+        pandas.DataFrame.plot : Draw plots.
```
You can get rid of `pandas.` and just leave `DataFrame.plot`. Something a bit more descriptive than "Draw plots." could be useful.

Not sure if in similar docstrings we also added the matplotlib equivalent method. Can you check the docstring for `DataFrame.kde` for example, and be consistent with it?

DZPM replied:
Added the `DataFrame.plot` description.

datapythonista commented on pandas/plotting/_core.py:
```diff
@@ -3337,19 +3337,87 @@ def kde(self, bw_method=None, ind=None, **kwds):
-        `**kwds` : optional
+        x : label or position, optional
+            Coordinates for the X axis. By default uses the index.
+        y : label or position, optional
```
As @dukebody says, for both `x` and `y`, in the first line we should have the type, more than a short description. It's a bit tricky, as `label` can be anything, but I think we're using something like `object or int, optional`. Then, we can sure clarify that it's the `label` or the `position` in the description.

DZPM replied:
I kept it consistent with  `DataFrame.plot`, which is the more complete docstring example I could find.

datapythonista commented on pandas/plotting/_core.py:
```diff
@@ -3337,19 +3337,87 @@ def kde(self, bw_method=None, ind=None, **kwds):
+            Coordinates for the X axis. By default uses the index.
+        y : label or position, optional
+            Column to plot. By default uses all columns.
+        stacked : boolean, default True
```
Happy to leave like this now, but as `stacked` is not in the signature, may be we should add it there, or document it under `kwds`. But if we leave it like this, can you replace `boolean` by the Python type `bool` please?

datapythonista reviewed (commented):
Looks great, just few minor things. Thanks for the changes!

DZPM committed:
More improvements

DZPM commented:
Changes done, and rebased. Hope it's ok!

TomAugspurger commented:
LGTM, thanks!

If you want, check the doc build when it finishes (probably a few hours).

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -3337,19 +3337,87 @@ def kde(self, bw_method=None, ind=None, **kwds):
+            ...     'visits': [20, 42, 28],
+            ...     'day': ['Monday', 'Tuesday', 'Wednesday'],
+            ... }, index=pd.date_range(start='2018/01/01', end='2018/07/01',
+            ...                        freq='M'))
```
This example is not running (the index is not correct). And also the labels to not show on the actual plot once this is fixed.

jorisvandenbossche reviewed (commented):
Some follow-up in https://github.com/pandas-dev/pandas/pull/22307
